### ISSUE TITLE
BUG overlapping particle geometry sometimes occurs

### ISSUE THREAD

**Issue by chrisk314** [chrisk314/mesh-sphere-packing#18]
Sometimes geometry of nearby particles overlaps which of course causes TetGen to fail. This problem only seems to show up when the characteristic segment length, `config.segment_length`, is small.

The attached config and particle input files lead to an overlap.

```
# config.yml
allow_overlaps: false
duplicate_particles: [true, false, false]
output_format: [multiflow]
particle_file: particles-p_629,cu_1.01,conc_0.567.msp
segment_length: 1.67e-05
tetgen_max_volume: 3.0e-13
tetgen_min_angle: 18.0
tetgen_rad_edge_ratio: 1.4
```

```
# particles.msp
4.213447766408998e-03 4.213287282590817e-03 4.213430842936750e-03
0 1 1
147 +3.335829707136979e-03 +1.235697961196803e-03 +2.250294937900698e-03 2.537845767162153e-04
580 +3.490679100750927e-03 +1.714022748431015e-03 +2.222754692406082e-03 2.505076711056732e-04
```

Here's an image from ParaView of the overlapping geometry.
![overlapping_particles](https://user-images.githubusercontent.com/2366658/37415372-30a0faf2-27a3-11e8-8b8f-47cb658a4b4f.png)

**Comment by chrisk314** [comment]
This issue occurs due to overlaps in the input data set which exceed the retraction amount which is [applied](https://github.com/chrisk314/mesh-sphere-packing/blob/master/mesh_sphere_packing/parse.py#L121-L122) to avoid overlapping geometry. It's time to do that `TODO`.

A sketch of the procedure would be as folllows:

1. Generate all sphere piece geometry
2. Insert all sphere pieces into a KD-tree in order to efficiently find particles which may overlap
3. If particle geometry overlaps then shrink the two overlapping hemispheres along the line of the branch vector connecting the particles to satisfy some imposed minimum separation distance.

Currently point 3. in the above is [applied](https://github.com/chrisk314/mesh-sphere-packing/blob/develop/mesh_sphere_packing/splitsphere.py#L387-L408) to particles near the walls. Ultimately the above procedure should be be generalised and applied to shrinking particles which have points very close to walls too.

**Comment by chrisk314** [comment]
Fixed by 4f10d54.

It would be nice to address the above point in future but for now this can be considered closed.
> Currently point 3. in the above is applied to particles near the walls. Ultimately the above procedure should be be generalised and applied to shrinking particles which have points very close to walls too.
